### ISSUE TITLE
http2Client.notifyError got notified that the client transport was broken EOF causes pubsub to stop working

### ISSUE THREAD

**Issue by Draiken** [GoogleCloudPlatform/google-cloud-go#556]
Hi there!

I have Go service that pulls pubsub messages and publishes them to some clients connected to it. After a period of time, normally some days, the service seems to just stop working. We've had problems reproducing the error since the logs would be rotated and we'd lose them. Finally I managed to get them.
This is our glide.lock: https://gist.github.com/Draiken/1d0036119658a9f77afc610584e4beab

We constantly see this error happening every 4 minutes.
```
http2Client.notifyError got notified that the client transport was broken EOF
```
We use GKE and have alerts on subscriptions that don't have messages pulled, but the metrics apparently tells us that the service is still pulling out data, even tho the only thing we see are these errors.

I need this service to error out, so Kubernetes can restart it if something is broken. I never thought about failing on message pulls since I expected some occasional failures, but once this error starts, it will continue happening for some time, and then the service goes radio silent. No log messages at all. I'm not an expert on Go, so my best guess is that the go routine stops.

We also have a second service that works in a similar way, pulling messages and sending them to other places, and it also stops working. Unfortunately I haven't gotten lucky with this one in order to catch the error, but I suspect it's suffering from the same problem.

Apart from trying to fix this error, I'd appreciate any directions I can take to identifying when this error happens and properly exiting the process.

Thanks

**Comment by hqiansc** [comment]
Looks like the same issue with https://github.com/grpc/grpc-go/issues/954

**Comment by jba** [comment]
The error message is annoying, but as far as I know it's just log spam. I didn't think it actually affected any services.

I just wrote a simple MessageIterator loop and ran it:
```
    s := pubsubClient.Subscription(subName)
    it, err := s.Pull(ctx)
    if err != nil {
        log.Fatal(err)
    }
    go func() {
        for {
            m, err := it.Next()
            if err != nil {
                log.Fatal(err)
            }
            log.Printf("ID %s: %s", m.ID, string(m.Data))
            m.Done(true)
        }
    }()
```
Then I used `gcloud beta pubsub topics publish` to send messages to it every once in a while. I observed the log spam, but it didn't stop my pull (the message numbering is an artifact of my test; it doesn't reflect publish order):
```
2017/03/17 10:51:00 ID 20750637330173: message-305
2017/03/17 10:51:02 ID 20750913597715: message-306
2017/03/17 10:51:03 ID 20750850458444: message-307
...
2017/03/17 11:46:21 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/03/17 11:46:47 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/03/17 11:47:11 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/03/17 11:47:49 ID 20753864721116: message-1
2017/03/17 11:47:50 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/03/17 11:47:50 ID 20753842012420: message-2
2017/03/17 11:47:51 ID 20753939609404: message-3
...
```
So in other words, I can't reproduce your problem. Could you isolate it down to a small program and post it here?

**Comment by Draiken** [comment]
Well, I couldn't reproduce the problem either. At least in a timely manner.
It just happens after days that the program is running as I said before.
I'll try to reproduce this again when I have time... you can close the
issue and I'll re-open if I manage to reproduce it successfully.

**Comment by jba** [comment]
OK. Sounds like your problem is unrelated to that log message.

**Comment by bradleyfalzon** [comment]
Btw Draiken, we've had this same issue, without a resolution, since adding more services to this queue it hasn't happened for a while though. I've always attributed this to some idle timeout after ~24 hours. I've just figured it was an underlying connection going away and nothing realised. I tracked it in https://github.com/bradleyfalzon/gopherci/issues/41 - but it's probably not very useful to you.

**Comment by absagar** [comment]
@jba my services seem to be working properly but my logs are flooded with these messages. Can I turn these logs somehow.

**Comment by adams-sarah** [comment]
Haven't tried this myself, but I think you should be able to set the grpc logger to at least log somewhere else. Or drop logs all together I suppose (though maybe not the best idea).
Then grpc wouldn't pollute stderr.

https://godoc.org/google.golang.org/grpc/grpclog

You should be able to do something in your init or main like:
```
import (
    "log"
    "google.golang.org/grpc/grpclog"
)

func init() {
    var myGRPCLogger grpclog.Logger = log.New(myIOWriter, "", log.LstdFlags)
    grpclog.SetLogger(myGRPCLogger)
}
```

**Comment by absagar** [comment]
yes, this is what I did as well. Thanks for the help @adams-sarah . Will wait till I get a cleaner fix.

**Comment by adams-sarah** [comment]
:+1:
